applications/Preprocessor/lib/elementShape.h

```diff
@@ -52,124 +57,118 @@ class ElementShape;
+///
+/// The ElementShape defines the shape of the possible reference elements in
+/// the mesh. Which are assumed to be polytopes (point, line, polygon,
+/// polyhedra). Such a reference polytope has several properties that are used
+/// in a FEM code:
+///
+///  1. A coordinate system, describing the geometry of the reference element
+///     and used for the basis functions.
+///  2. The topology of the reference element, as a polytope. That is how are
+///     the parts (vertices, edges, etc.) connected to form the polytope.
+///  3. A numbering of the boundary parts so that each part can be indexed and
+///     their orientation.
+/// In the kernel we don't need the geometry part, but we do need the topology
+/// and the indices of the parts.
+///
+/// To define the topology of the polytope we use two properties:
+///   1. It's boundary is a combination of 0, 1, ... D-1, dimensional polytopes
+///   2. It has N vertices (0 dimensional polytopes).
+/// The vertices are numbered 0...N-1. The intermediate polytopes on the
+/// boundary are then defined by taking a subset of the vertices and providing
+/// the associated polytopical shape. The ordering of the vertices then defines
+/// the relative orientation.
+///
+/// Note: As both of these properties influence the output, they should match
+/// the definition of the shapes in the kernel.
+///
+/// \tparam dimension The dimension of this shape
 template <std::size_t dimension>
 class ElementShape {
-    template <std::size_t d>
-    struct tag {};
 
    public:
     ElementShape() = default;
 
     ~ElementShape() = default;
 
-    ElementShape(const ElementShape&) = default;
+    // Disallow copying and moving for two reasons:
+    // 1. It contains a self reference as part of its shapeParts, this would
+    //    need a custom copying and moving routine to keep this reference
+    //    correct
+    // 2. Larger shapes store pointers to smaller shapes (for the boundary).
+    //    By moving and copying one could easily break these pointers.
 
-    ElementShape(ElementShape&&) noexcept = default;
+    ElementShape(const ElementShape&) = delete;
 
-    ElementShape& operator=(const ElementShape&) = default;
+    ElementShape(ElementShape&&) noexcept = delete;
 
-    ElementShape& operator=(ElementShape&&) noexcept = default;
+    ElementShape& operator=(const ElementShape&) = delete;
+
+    ElementShape& operator=(ElementShape&&) noexcept = delete;
 
     template <typename... Args>
-    ElementShape(Args... args) : shapeData(args...) {
-        completeSubShapes(tag<dimension - 1>{}, tag<dimension - 1>{});
+    ElementShape(Args... args) : shapeParts({}, args...) {
+        // Insert ElementShapePart corresponding to the whole element.
+        auto& shapeVec = shapeParts.template get<dimension>();
+        std::vector<std::size_t> vertices(getNumberOfNodes());
+        std::iota(vertices.begin(), vertices.end(), 0);
+        shapeVec.emplace_back(
```

Comment:
can be a push_back I think

Reply:
Yes, or I should remove the constructor.

Reply:
Updated